flink-connector-cassandra/src/main/java/org/apache/flink/connector/cassandra/source/split/CassandraSplit.java

```diff
@@ -0,0 +1,66 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.connector.cassandra.source.split;
+
+import org.apache.flink.api.connector.source.SourceSplit;
+
+import java.io.Serializable;
+import java.util.Objects;
+import java.util.Set;
+
+/**
+ * {@link SourceSplit} for Cassandra source. A Cassandra split is just a set of {@link RingRange}s
+ * (a range between 2 tokens). Tokens are spread across the Cassandra cluster with each node
+ * managing a share of the token ring. Each split can contain several token ranges in order to
+ * reduce the overhead on Cassandra vnodes.
+ */
+public class CassandraSplit implements SourceSplit, Serializable {
+
+    private final Set<RingRange> ringRanges;
+
+    public CassandraSplit(Set<RingRange> ringRanges) {
+        this.ringRanges = ringRanges;
+    }
+
+    public Set<RingRange> getRingRanges() {
+        return ringRanges;
+    }
+
+    @Override
+    public String splitId() {
+        return ringRanges.toString();
```

Comment:
Well, `ringRanges` are modified only in the `CassandraSplit` constructor so they are immutable. But the problem could be more on `CassandraSplitState` : its `unprocessedRingRanges` are mutated during the reading of splits. But `CassandraSplitState` is initialized from the `CassandraSplit`s received by the enumerator with `CassandraSplit`'s `splitId` .  `CassandraSplitState`  also serves to restore splits in case of failure and create the corresponding `CassandraSplit` reusing its `splitId`.  So I don't think there is any possible dynamic changing of a `splitId`.